src/dataflow/src/server/storage_state.rs

```diff
@@ -0,0 +1,311 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+
+use std::collections::HashMap;
+use std::rc::Weak;
+use std::time::Instant;
+
+use timely::communication::Allocate;
+use timely::order::PartialOrder;
+use timely::progress::frontier::Antichain;
+use timely::progress::ChangeBatch;
+use timely::worker::Worker as TimelyWorker;
+use tokio::sync::mpsc;
+
+use mz_dataflow_types::client::{
+    Response, StorageCommand, StorageResponse, TimestampBindingFeedback,
+};
+use mz_dataflow_types::sources::{ExternalSourceConnector, SourceConnector};
+use mz_expr::{GlobalId, PartitionId};
+use mz_ore::now::NowFn;
+use mz_persist::client::RuntimeClient;
+use mz_repr::Timestamp;
+
+use crate::metrics::Metrics;
+use crate::render::sources::PersistedSourceManager;
+use crate::server::LocalInput;
+use crate::source::metrics::SourceBaseMetrics;
+use crate::source::timestamp::TimestampBindingRc;
+use crate::source::SourceToken;
+
+/// How frequently each dataflow worker sends timestamp binding updates
+/// back to the coordinator.
+static TS_BINDING_FEEDBACK_INTERVAL_MS: u128 = 1_000;
```

Comment:
`Duration` constructors are `const` since Rust 1.58, so we could change the type to `const Duration` here. Can do in a separate PR.

Reply:
Cool. I just copy/pasted but happy to adapt / have someone else adapt.